### ISSUE TITLE
Compile errors with latest strictest TypeScript

### ISSUE THREAD

**Issue by Stouffi** [brillout/vite-plugin-ssr#145]
I am struggling getting my project compiled successfully using `vite-plugin-ssr`.
I suspect there is some issue with the published package that makes the typescript compiler resolve wrong module when using imports from `vite-plugin-ssr/client`
The compiler will resolve with `<a-long-path-to>.zip/node_modules/vite-plugin-ssr/client/index`, that is the `index.ts` file instead of a proper `index.d.ts` file.
This makes part of the library not skippable using the `skipLibCheck` option that is so useful when one uses the strictest compiler options available for their own source files even if some libraries are not as strict.
I could get it to work with the help of the paths compiler option

```json
"paths": {
  "vite-plugin-ssr/client": [
    "<the-long-path-to>.zip/node_modules/vite-plugin-ssr/dist/esm/client"
  ]
}
```

But this can only be a temporary workaround.

After some searches, I found an interesting issue https://github.com/microsoft/TypeScript/issues/33079 which subject is the lack of support for the exports field in package.json.

There are at least 2 workarounds:

https://github.com/microsoft/TypeScript/issues/33079#issuecomment-738033267
This one requires that the published package matches its directory structure 1:1 with import paths.

Another workaround below:
https://github.com/microsoft/TypeScript/issues/33079#issuecomment-757423622
This one requires the addition of minimal package.json in top-level directories that matches 1:1 with import paths.

I'm unfortunately not experienced in how to change things related to package publishing. I hope at least that the above information will help for a fix.

**Comment by brillout** [comment]
I'm aware of this issue.

> latest strictest TypeScript

Which are?

An alternative workaround is to make `vite-plugin-ssr` comply to the strictest settings.

**Comment by brillout** [comment]
Note that the boilerplates compile just fine (e.g. https://github.com/brillout/vite-plugin-ssr/blob/master/boilerplates/boilerplate-react-ts/tsconfig.json) and these use `"typescript": "^4.4.2"` with `"strict": true`. So I'm curious what are your tsconfig settings that makes vite-plugin-ssr choke.

**Comment by Stouffi** [comment]
I'm using these additional checks:

```
"noUnusedLocals": true, /* Enable error reporting when a local variables aren't read. */
"noUnusedParameters": true, /* Raise an error when a function parameter isn't read */
"exactOptionalPropertyTypes": true, /* Interpret optional property types as written, rather than adding 'undefined'. */
"noImplicitReturns": true, /* Enable error reporting for codepaths that do not explicitly return in a function. */
"noFallthroughCasesInSwitch": true, /* Enable error reporting for fallthrough cases in switch statements. */
"noUncheckedIndexedAccess": true, /* Include 'undefined' in index signature results */
"noImplicitOverride": true, /* Ensure overriding members in derived classes are marked with an override modifier. */
"noPropertyAccessFromIndexSignature": true
```
All those checks are not (yet) implied by `strict: true`

But definitely `vite-plugin-ssr` should be published in a way that makes the TS compiler resolving the modules import to their corresponding type declarations, that's how a library is supposed to be consumed.
This way the consumer can work using any set of checks regardless of what was used to build the library.

e.g.

```typescript
import { html } from 'vite-plugin-ssr/client' // 'vite-plugin-ssr/client' should 
// be resolved as "<the-long-path-to>.zip/node_modules/vite-plugin-ssr/dist/esm/client/index"
```

You might think that all these checks are cumbersome :slightly_smiling_face:, that's why, imho, it'd be more interesting to fix the publish rather than complying with settings that might changed at the next minor release of typescript.

**Comment by brillout** [comment]
> You might think that all these checks are cumbersome

On the contrary; I was actually looking for a more strict setup but was too lazy to hunt down all settings. I added all your settings to `vite-plugin-ssr`'s tsconfig.

> fix the publish rather than complying with settings

I'd rather wait for TS to supports `exports`. From https://github.com/microsoft/TypeScript/issues/44522#issuecomment-861528873:

> Maybe 4.4 will get it, little bit hard to tell yet as there's a lot to cover with node's new ESM resolvers #44442

So the hope is that TS will add support soon. If this doesn't happen then yea we'll have to dig some workaround like the ones you suggested.

Until then I'm more than up for adding more strict settings to the plugin's tsconfig.

I will release the new version tomorrow.

**Comment by brillout** [comment]
Tmrw. (Going to be released with a major semver bump and I want to include further breaking changes.)

**Comment by brillout** [comment]
Released in `0.3.1`.

Closing but we can re-open if we encounter TS problems again.

Hopefully TS will support the `exports` field sooner rather than later.
